### PR TITLE
Initial JSON support.

### PR DESCRIPTION
**Pull request by tcf909**
Made sure to support JSON strings as values (not just objects) so that someone could stringify their own objects (and avoid risking post execution changes to an object they provided to be serialized).

Still need to add test cases. Getting a PR started to start the discussion.

### CODE REVIEW

**Comment by mbautin** [comment]
Hi @tcf909, thank you very much for your contribution! Would you mind signing the CLA (Contributor License Agreement) using this form? https://docs.google.com/forms/d/11hn-vBGhOZRunclC3NKmSX1cvQVrU--r0ldDLqasRIo

**Comment by tcf909** [comment]
> Hi @tcf909, thank you very much for your contribution! Would you mind signing the CLA (Contributor License Agreement) using this form? https://docs.google.com/forms/d/11hn-vBGhOZRunclC3NKmSX1cvQVrU--r0ldDLqasRIo

Done

**Comment by mbautin** [comment]
@tcf909 thanks a lot! We are looking forward to reviewing and accepting your patch.

**Comment by tcf909** [comment]
@mbautin Happy to help. Didn't seem like a lot was needed on the client side. Let me know if I missed anything.

**Comment by tcf909** [comment]
I did miss a couple of things. Added the guess functionality as well as the needed tests.

A couple of notes: Right now the driver absolutely will not encode a value === null. This is very early in the encode process. Someone should decide if they would like to change this default behavior.

A couple of other notes:

Arrays are still guessed as being a List and a boolean is guessed as being a boolean (not JSON). I didn't change these historic guesses.

**Review by andrewc-dev — COMMENTED** [review]
Can you add a test case for when the object contains functions as values?
`{ a: 1, b: function () { return 'foo' } }` becomes `'{"a":1}'` when JSON.stringify is called.

**Review by andrewc-dev — APPROVED** [review]
LG!

**Comment by tcf909** [comment]
> Can you add a test case for when the object contains functions as values?
> `{ a: 1, b: function () { return 'foo' } }` becomes `'{"a":1}'` when JSON.stringify is called.

Wouldn't that be desired behavior? Are you looking to serialize functions?

**Comment by andrewc-dev** [comment]
Yes, either we change the function to a string or remove the value. Like you said, I think this is the desired behavior so adding it for more complete test coverage couldn't hurt.

**Comment by tcf909** [comment]
> Yes, either we change the function to a string or remove the value. Like you said, I think this is the desired behavior so adding it for more complete test coverage couldn't hurt.

I think the upstream testing on JSON.stringify (which happens in the v8 project) is what you want to rely on here isn't it? Otherwise you need to go through the 30 or so v8 complex objects (REGEXP, DATE, etc..). Also you would need to test for .toJSON() edge cases. There are so many cases that a user can trigger when we are talking JSON in JS that I'm not sure how you tackle locking down the edge cases in a more holistic sense.

I've gotten rid of the fork of this project, so I'd need to go back in re-apply this pr to make the change.

Feel free to modify the PR as desired or take the code in to an alternative PR.

**Comment by andrewc-dev** [comment]
You are right, that seems beyond the scope of this driver. Thanks for the insight!

File: lib/encoder.js [hunk 1]
```diff
@@ -752,6 +755,31 @@ function defineInstanceMembers() {
...
   };
+  this.encodeJson = function encodeJson(value){
+    let jsonString;
+    //Use a switch to handle other typeof checking in the future
+    switch(typeof value){
+      case 'string':
+        // If valid json was already provided (but stringified) use the provided string
+        // This is valid if a user wants absolute control over the json object they pass for storage
+        // Passing an object as a value would leave some risk of the object being modified by the stack after client.execute() is called
+        try{
```
Comment (tcf909) [inline lib/encoder.js:766]: This try/catch block should actually be removed so malformed json strings do not make it past this point.

File: lib/encoder.js [hunk 2]
```diff
@@ -752,6 +755,31 @@ function defineInstanceMembers() {
     parts.unshift(this.getLengthBuffer(propCounter));
     return Buffer.concat(parts);
   };
+  this.encodeJson = function encodeJson(value){
+    let jsonString;
+    //Use a switch to handle other typeof checking in the future
+    switch(typeof value){
+      case 'string':
```
Comment (tcf909) [inline lib/encoder.js:762]: The switch statement should be converted to an "if string else json" statement until there is a need for multiple type checking.

**Comment by rkarthik007** [comment]
Super happy to see this go in! @tcf909 - could you please add yourself to the yb contributors list here: https://github.com/YugaByte/yugabyte-db/blob/master/CONTRIBUTORS.md

You would need to send a PR, I can merge it in.

**Review by tcf909 — COMMENTED** [review]
1 serious issue that could lead to malformed JSON making it past the driver, and 1 minor optimization to save building a switch statement.

**Comment by tcf909** [comment]
> Super happy to see this go in! @tcf909 - could you please add yourself to the yb contributors list here: https://github.com/YugaByte/yugabyte-db/blob/master/CONTRIBUTORS.md
> 
> You would need to send a PR, I can merge it in.

Thanks Karthik. No need to pollute your list with my small contribution. Hope you guys are doing well.

File: lib/encoder.js (hunk 1 above)
Comment (tcf909) [reply lib/encoder.js:766]: The code in the try/catch statement should still stay though. This will just allow JSON.parse to throw if the string is malformed (just like the JSON.stringify would in the default: switch statement)

**Comment by rkarthik007** [comment]
> > Super happy to see this go in! @tcf909 - could you please add yourself to the yb contributors list here: https://github.com/YugaByte/yugabyte-db/blob/master/CONTRIBUTORS.md
> > You would need to send a PR, I can merge it in.
> 
> Thanks Karthik. No need to pollute your list with my small contribution. Hope you guys are doing well.

We're doing well, hope you are too! This is actually becoming a much requested feature, so happy to have you listed, but no issues otherwise.
